File: apps/mobile/src/lib/plan-client.ts
import {
  getActivePlan as bridgeGetActivePlan,
  savePlan as bridgeSavePlan,
  applySavedPlanToToday as bridgeApplySavedPlanToToday,
  type GrowthModeKey,
  type GrowthRule,
  type PlanEditInput,
  type PlanSummary,
} from './mobile-bridge';

export type {PlanSummary, PlanEditInput};

const GROWTH_MODE_KEYS: GrowthModeKey[] = [
  'newWord',
  'review',
  'mixedTest',
  'wrongWordReinforcement',
  'rootAffix',
];

export function normalizeGrowthRule(rule?: Partial<GrowthRule> | null): GrowthRule {
  return {
    intervalDays: Math.max(1, Math.min(365, rule?.intervalDays ?? 7)),
    increment: Math.max(0, Math.min(100, rule?.increment ?? 5)),
  };
}

export function normalizePlanGrowth(plan: PlanSummary): PlanSummary {
  const sharedGrowthRule = normalizeGrowthRule(
    plan.sharedGrowthRule ?? {
      intervalDays: plan.growthIntervalDays,
      increment: plan.growthIncrement,
    },
  );
  const growthRulesByMode = Object.fromEntries(
    GROWTH_MODE_KEYS.map(mode => [
      mode,
      normalizeGrowthRule(plan.growthRulesByMode?.[mode] ?? sharedGrowthRule),
    ]),
  ) as Record<GrowthModeKey, GrowthRule>;

  return {
    ...plan,
    growthRuleMode: plan.growthRuleMode ?? 'shared',
    growthIntervalDays: sharedGrowthRule.intervalDays,
    growthIncrement: sharedGrowthRule.increment,
    sharedGrowthRule,
    growthRulesByMode,
  };
}

export async function fetchActivePlan(): Promise<PlanSummary | null> {
  const plan = await bridgeGetActivePlan();
  return plan ? normalizePlanGrowth(plan) : null;
}

export async function savePlan(
  planId: number,
  input: PlanEditInput,
): Promise<PlanSummary> {
  const saved = await bridgeSavePlan(planId, input);
  return normalizePlanGrowth(saved);
}

export async function applyPlanToToday(): Promise<PlanSummary> {
  const applied = await bridgeApplySavedPlanToToday();
  return normalizePlanGrowth(applied);
}

export function validatePlanInput(input: PlanEditInput): string | null {
  if (
    input.newWordsPerDay !== undefined &&
    (input.newWordsPerDay < 0 || input.newWordsPerDay > 100)
  ) {
    return '每日新词数量需要在 0 到 100 之间';
  }

  if (
    input.reviewWordsPerDay !== undefined &&
    (input.reviewWordsPerDay < 0 || input.reviewWordsPerDay > 200)
  ) {
    return '每日复习数量需要在 0 到 200 之间';
  }

  if (
    input.mixedTestPerDay !== undefined &&
    (input.mixedTestPerDay < 0 || input.mixedTestPerDay > 50)
  ) {
    return '每日混合测试数量需要在 0 到 50 之间';
  }

  if (
    input.wrongWordTestPerDay !== undefined &&
    (input.wrongWordTestPerDay < 0 || input.wrongWordTestPerDay > 30)
  ) {
    return '每日错词强化数量需要在 0 到 30 之间';
  }

  if (
    input.rootAffixPerDay !== undefined &&
    (input.rootAffixPerDay < 0 || input.rootAffixPerDay > 50)
  ) {
    return '每日词根词缀数量需要在 0 到 50 之间';
  }

  if (
    input.growthIntervalDays !== undefined &&
    (input.growthIntervalDays < 1 || input.growthIntervalDays > 365)
  ) {
    return '增长周期需要在 1 到 365 天之间';
  }

  if (
    input.growthIncrement !== undefined &&
    (input.growthIncrement < 0 || input.growthIncrement > 100)
  ) {
    return '增长幅度需要在 0 到 100 之间';
  }

  if (input.sharedGrowthRule) {
    const shared = normalizeGrowthRule(input.sharedGrowthRule);
    if (shared.intervalDays < 1 || shared.intervalDays > 365) {
      return '共享增长周期需要在 1 到 365 天之间';
    }
    if (shared.increment < 0 || shared.increment > 100) {
      return '共享增长幅度需要在 0 到 100 之间';
    }
  }

  if (input.growthRulesByMode) {
    for (const mode of Object.keys(input.growthRulesByMode) as GrowthModeKey[]) {
      const rule = normalizeGrowthRule(input.growthRulesByMode[mode]);
      if (rule.intervalDays < 1 || rule.intervalDays > 365) {
        return `${mode} 的增长周期需要在 1 到 365 天之间`;
      }
      if (rule.increment < 0 || rule.increment > 100) {
        return `${mode} 的增长幅度需要在 0 到 100 之间`;
      }
    }
  }

  return null;
}

File: apps/flutter_mobile/lib/sdk/plan_client.dart
/// Plan client - typed SDK for plan and wordbook APIs.
library;

import '../bridge/bridge.dart';

class PlanSummary {
  final int id;
  final String name;
  final int newWordsPerDay;
  final int reviewWordsPerDay;
  final int mixedTestPerDay;
  final int wrongWordTestPerDay;
  final int? rootAffixPerDay;
  final int growthIntervalDays;
  final int growthIncrement;
  final String growthRuleMode;
  final Map<String, dynamic>? sharedGrowthRule;
  final Map<String, dynamic>? growthRulesByMode;
  final Map<String, dynamic>? questionTypeWeightsByMode;

  const PlanSummary({
    required this.id,
    required this.name,
    required this.newWordsPerDay,
    required this.reviewWordsPerDay,
    required this.mixedTestPerDay,
    required this.wrongWordTestPerDay,
    this.rootAffixPerDay,
    required this.growthIntervalDays,
    required this.growthIncrement,
    required this.growthRuleMode,
    this.sharedGrowthRule,
    this.growthRulesByMode,
    this.questionTypeWeightsByMode,
  });

  factory PlanSummary.fromJson(Map<String, dynamic> json) {
    final sharedRule =
        (json['sharedGrowthRule'] as Map?)?.cast<String, dynamic>() ??
        <String, dynamic>{
          'intervalDays': json['growthIntervalDays'] as int? ?? 7,
          'increment': json['growthIncrement'] as int? ?? 5,
        };
    final mode = (json['growthRuleMode'] as String?) ?? 'shared';
    final modes = <String, dynamic>{
      ...?((json['growthRulesByMode'] as Map?)?.cast<String, dynamic>()),
    };
    for (final key in const [
      'newWord',
      'review',
      'mixedTest',
      'wrongWordReinforcement',
      'rootAffix',
    ]) {
      modes[key] ??= Map<String, dynamic>.from(sharedRule);
    }

    return PlanSummary(
      id: json['id'] as int,
      name: json['name'] as String,
      newWordsPerDay: json['newWordsPerDay'] as int,
      reviewWordsPerDay: json['reviewWordsPerDay'] as int,
      mixedTestPerDay: json['mixedTestPerDay'] as int,
      wrongWordTestPerDay: json['wrongWordTestPerDay'] as int,
      rootAffixPerDay: json['rootAffixPerDay'] as int?,
      growthIntervalDays:
          (sharedRule['intervalDays'] as num?)?.toInt() ??
          (json['growthIntervalDays'] as int? ?? 7),
      growthIncrement:
          (sharedRule['increment'] as num?)?.toInt() ??
          (json['growthIncrement'] as int? ?? 5),
      growthRuleMode: mode,
      sharedGrowthRule: sharedRule,
      growthRulesByMode: modes,
      questionTypeWeightsByMode:
          (json['questionTypeWeightsByMode'] as Map?)?.cast<String, dynamic>(),
    );
  }
}

class WordbookSummary {
  final int id;
  final String code;
  final String name;
  final String category;
  final int totalEntries;
  final bool isActive;

  const WordbookSummary({
    required this.id,
    required this.code,
    required this.name,
    required this.category,
    required this.totalEntries,
    required this.isActive,
  });

  factory WordbookSummary.fromJson(Map<String, dynamic> json) => WordbookSummary(
        id: json['id'] as int,
        code: json['code'] as String,
        name: json['name'] as String,
        category: json['category'] as String,
        totalEntries: json['totalEntries'] as int,
        isActive: json['isActive'] as bool,
      );
}

class PlanClient {
  final RustBridge _bridge;
  final BridgeCodec _codec;

  const PlanClient(this._bridge, this._codec);

  Future<PlanSummary?> getActivePlan() async {
    final raw = await _bridge.call('getActivePlan');
    final json = _codec.decodeResponse(raw);
    if (json.isEmpty) return null;
    return PlanSummary.fromJson(json);
  }

  Future<PlanSummary> savePlan({
    required int planId,
    required Map<String, dynamic> input,
  }) async {
    final request = <String, dynamic>{'planId': planId, 'input': input};
    final raw = await _bridge.call('savePlan', _codec.encodeRequest(request));
    final json = _codec.decodeResponse(raw);
    return PlanSummary.fromJson(json);
  }

  Future<PlanSummary> applySavedPlanToToday() async {
    final raw = await _bridge.call('applySavedPlanToToday');
    final json = _codec.decodeResponse(raw);
    return PlanSummary.fromJson(json);
  }

  Future<List<WordbookSummary>> getWordbooks() async {
    final raw = await _bridge.call('getWordbooks');
    final decoded = _codec.decodeDynamicResponse(raw);
    if (decoded is! List) return const [];
    return decoded
        .whereType<Map<String, dynamic>>()
        .map(WordbookSummary.fromJson)
        .toList(growable: false);
  }

  Future<void> toggleWordbook({
    required int wordbookId,
    required bool isActive,
  }) async {
    await _bridge.callVoid(
      'toggleWordbook',
      _codec.encodeRequest({'wordbookId': wordbookId, 'isActive': isActive}),
    );
  }
}
